### DISCUSSION TITLE
Inference speed on the Google Colab notebook provided by default

### DISCUSSION THREAD

**Discussion by oliviermirat** [talmolab/sleap#1276] [Help!] [answered]
Hello,

After testing the Google Colab notebook provided by default (https://colab.research.google.com/github/talmolab/sleap/blob/main/docs/notebooks/Training_and_inference_on_an_example_dataset.ipynb) I saw that the inference time of the network is about 10 fps (even with the paid version of Google Colab with better GPUs). In the 2022 paper, I see that inference times can go up to 800 fps. I would like to understand why there's such a big difference (is this related to the way the video is loaded? or to some model parameters?) and how I could make the inference run faster on Google Colab and/or on a local machine with GPU. 

Thanks in advance for your help!

Olivier

**Comment by roomrys** [comment]
Hi @oliviermirat,

Some parameters that will affect speed (at the expense of memory usage):
1. Hardware: of course, the compute power of the CPU/GPU matters. Upgrading to Colab Pro likely gives you the max compute power for an on-the-cloud GPU, so let's look at some others...
2. Batch size: This is the number of images the model processes at once. A larger batch size increases the fps since the model is processing more data in parallel, but requires more memory usage.
3. Image size: Larger images require more processing power, slowing down the model.
4. Model Complexity: There are few hyperparameters that deal with model complexity, namely the number of layers (affected by *max stride*) and the number of *filters* (and the *filter rate*). More complex models (having more layers and more filters) can slow down inference.

### Recommendations
- If you have enough memory, then you can try increasing your batch size. The annoying thing about running out of memory is that Colab will just exit at some point during inference after having run for a while (without any error message). To check if you ran out of memory, you need to monitor the resources
![image](https://user-images.githubusercontent.com/38435167/232568289-9ed44795-1985-49f0-8c63-e060a5e7a54b.png)

- If you are tracking large features (as opposed to small features that are only ~2 pixels large), then you can set the input scaling lower (effectively reducing the image size into the network).
- If your model is working fine, then I would leave this step as a last resort since we will directly be affecting the model's ability to find features. But, you can start decreasing the max stride and/or the number of filters/filter rate. If you notice that model performance is going down, then I would revert to the lowest settings for stride/filters that works as expected.

Thanks,
Liezl

**Reply by oliviermirat** [reply]
Thanks a lot for your detailed answer Liezl!

I've tried changing the batch size, max stride and number of filters but I still get an inference speed of about 10fps. The video I'm using is the one provided by default with the software (with the 2 flies, resolution 1024*1024) (which I'm guessing is the one you used to get 800fps in the paper?), so I haven't tried changing the image size since the paper specifies that no down-sampling of the images had been done.

I'm using this command on google colab to test the inference speed of the fps :
!sleap-track colab_demo.mp4
--frames 0-200
--tracking.tracker simple
-m "models/230418_152348.centered_instance"
-m "models/230418_152348.centroid"
(the other option to test inference doesn't work for some unknown reason)
Could I be using the wrong command? Maybe the fact that it's reading from the video that's on Google Drive is causing this 10fps inference speed?

Thanks a lot for your help,

Olivier

**Reply by talmo** [reply, marked as answer]
Hi @oliviermirat,

A few more factors:

1. This command includes I/O, meaning the time it takes to load the video and decode the frames. This can vary significantly based on the CPU and, most importantly, where the video is being read from, which in this case will be slow drives on the cloud VMs that Colab notebooks run on.
2. This command also includes tracking, which is considerably slower than model inference. This is not the case in supervised ID models, which give you identity and pose at the same time.
3. The numbers in the paper were achieved using TensorRT for acceleration on an RTX Titan with FP16 quantization at batch size 16. [See this repo for more info on TensorRT](https://github.com/sheridana/sleap_tensor_rt).

**Reply by oliviermirat** [reply]
Hi @talmo 

Thanks a lot for pointing out these additional factors! 

I think the next step for me will be to experiment on a local computer or on a cluster/server, probably with the supervised ID models and potentially also with the TensorRT model.

Thanks a lot!

Olivier
